### PR TITLE
[CompilerPerf] Weak ByteFile[]

### PR DESCRIPTION
**Pull request by dsyme**
This is a possible solution to the "ByteFile" memory usage problem discussed [here](https://github.com/Microsoft/visualfsharp/pull/4429#issuecomment-372662705) and [here](https://github.com/Microsoft/visualfsharp/issues/4258)

### Background

ByteFile accounts for something like 25% of F# memory usage in Visual Studio and long running FCS applications (we have to revalidate that this is still the case, but we expect it to be)

@vbfox says:

> From my measurements ByteFile is the single thing that make VS OOM for our team.
> Both the compiler directly and via type providers maintain them in memory, and having a dll with some bitmap resources of a few dozen MB can trash VS memory very fast.
> And from what I saw in the code only the tables (types, methods, ...) are used and not the IL or the resources or any other.

ByteFile objects come from DLL references - we read the whole DLL into memory. The memory gets collected if you move on to a new project or the builder for the project is otherwise collected.  But while the project is being edited the memory is held strongly.  

The memory held by `byte[]` is not very intrusive for GC, but is a killer when the memory reaches max address space, which affects Visual Studio in particular (most other F# editing tools use 64-bit address spaces)

### Proposed Solution

The solution here is to hold the ByteFile memory weakly for stable DLLs.  Stable DLLs can be whatever we choose them to be, but I have chosen them in this PR to be:

* Stable DLL = DLLs whose full path name contains "packages" or "Reference Assemblies" or "lib/mono"

The prototype does what it says on the tin: it holds the bytes weakly, resurrects them if they are needed and have been collected, and complains if the file has been written to under its feet.

The PR is carefully written to hold the contents strongly over  operations in ilread.fs that eagerly read data chunks.  It does this by converting the WeakByteFile into a ByteFile for the duration of an operation.

The prototype should only change the behaviour of FCS, Visual Studio and F# Interactive.  It doesn't apply to fsc.exe which generally uses memory mapping for these binary files.

Done:

* [x] Get it green
* [x] Validate via a memory dump that we have reduced active byte[] usage

### CODE REVIEW

**Comment by smoothdeveloper** [comment]
Linking that discussion/issue as well #2345 as it was discussing the matter of ByteFile and concern with assembly resources being held around through it.

Can we make stable dlls discard dlls that contains large amount of resources? I don't think that the current criteria will help in the situation I'm facing.

**Comment by dsyme** [comment]
> Can we make stable dlls discard dlls that contains large amount of resources? I don't think that the current criteria will help in the situation I'm facing.

We could.  Or we could just change the reader to only read the metadata section.  Or ByteFile to read in chunks.

**Comment by dsyme** [comment]
Scenario: 
   1. open FSharp.sln + close all files + exit + open FSharp.sln again
   2. open IlxGen.fs, go to end of file, Find All Symbols on `CreatePermissionSets`, wait for symbols and low CPU, take dump

Memory after this PR:

image

Memory by type in dev15.6:

image

**Comment by dsyme** [comment]
@AviAvni Notes:

* dev15.6 = 109MB managed memory heap
* This PR = 96MB managed memory heap

That's a 12% overall reduction from the combined work.  It's a slightly inexact science though and this scenario is favourable to dev15.6 in some ways I think

* ByteFile has disappeared as a major source (though there is still a 3MB non-stable ByteFile for a reference to FSharp.Core?) 

* Good to see the Val data being halved

Future work:

* There are many more `Typar` than I realised (10%).  We should do reduction work on that after all

* `TcSymbolUses` (11%) is high after find-all-symbols, I suppose that's to be expected and the data will be cleaned up later

* `FSharpTokenInfo` is now appearing (5%), it must be being held by the colorizer

image

* There are also lists of FSharpTokenInfo held in `SourceLineData` (2%), I think better is possible here

image

* These lists of `open` declarations (3%) being held as part of the results of background analysis are taking more memory than I would expect.  These could be RessizeArray's, for example, and perhaps there are duplicates in the list

image

* The file checking cache in `backgroundCompiler` holds a string (2%) which is the entire contents of each file, perhaps we can just hold a hash instead.  The strings are held in this tuple:

image

**Comment by AviAvni** [comment]
@dsyme thanks starting to work on Typar
the field to move to the OptionalData are: typar_il_name, typar_xmldoc, typar_attribs, typar_solution, typar_constraints
is it ok?

**Comment by dsyme** [comment]
@AviAvni Not quite. The common case is for **inference** type variables.  So these are the potential optional ones:

* typar_il_name
* typar_xmldoc
* typar_attribs
* typar_constraints

You _might_ be able to do flags as well, see https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/tast.fs#L1920

But I'm not sure how the stats will work out - we should be data-driven about this...

**Comment by dsyme** [comment]
@KevinRansom @TIHan @brettfo This is ready - it definitely reduces memory usage for VS, and I've not found any problems with it. Please review, thanks

File: src/absil/ilread.fs [hunk 1]
```diff
@@ -245,6 +257,93 @@ type ByteFile(bytes:byte[]) =
...
         uint16 b0 ||| (uint16 b1 <<< 8) 
+
+    override bf.HoldEager = (bf :> BinaryFile)
+
+type WeakBytes(filename) = 
+#if DEBUG
+    static let mutable created = 0
+    static let mutable recreated = 0
+    static let mutable all : WeakBytes list = []
+    static do AppDomain.CurrentDomain.ProcessExit.Add(fun _ ->
```
Comment (KevinRansom) [inline src/absil/ilread.fs]: AppDomain is not coreclr, so this won't work when debugging on the coreclr

File: src/fsharp/service/IncrementalBuild.fs [hunk 2]
```diff
@@ -1726,10 +1726,6 @@ type IncrementalBuilder(tcGlobals, frameworkTcImports, nonFrameworkAssemblyInput
                 // see also fsc.fs:runFromCommandLineToImportingAssemblies(), as there are many similarities to where the PS creates a tcConfigB
                 let tcConfigB = TcConfigBuilder.CreateNew(legacyReferenceResolver, defaultFSharpBinariesDir, implicitIncludeDir=projectDirectory, optimizeForMemory=true, isInteractive=false, isInvalidationSupported=true, defaultCopyFSharpCore=false) 
 
-                // The following uses more memory but means we don't take read-exclusions on the DLLs we reference 
-                // Could detect well-known assemblies--ie System.dll--and open them with read-locks 
-                tcConfigB.openBinariesInMemory <- true
-
                 tcConfigB.resolutionEnvironment <- (ReferenceResolver.ResolutionEnvironment.EditingOrCompilation true)
```
Comment (KevinRansom) [inline src/fsharp/service/IncrementalBuild.fs]: Is optimizeForMemory not required to be set here?

File: vsintegration/vsintegration.targets [hunk 3]
```diff
@@ -1,7 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
```
Comment (KevinRansom) [inline vsintegration/vsintegration.targets]: Is this dead code?

File: src/fsharp/service/IncrementalBuild.fs (hunk 2 above)
Comment (dsyme) [reply src/fsharp/service/IncrementalBuild.fs]: It's set on the previous line.  There was duplication between `openBinariesInMemory` and `optimizeForMemory`, best just to remove the first and use the second.

**Comment by dsyme** [comment]
I've received some feedback on this from chattting to the Roslyn team

1. Like Visual F#, Roslyn uses an engineered solution to this problem that's different for VS to the command-line compiler.  In Roslyn VS, there is a sophisticated manager of managed DLL memory resources that uses views on a memory-mapped file to reduce memory usage.  

2. We could consider passing a way to access this infrastructure into FCS as part of its configuration.  That would have a major advantage that C# and F# share these resources in a VS instance.  However not the whole part of the Roslyn API is public, but perhaps enough of it is.  

3. The Roslyn solution does achieve the "opening the binary in-memory never goes back to the file nor re-reads from the file" result we're after

**Comment by dsyme** [comment]
Closed in favour of #4586

**Comment by dsyme** [comment]
@KevinRansom 

> Throughout this file is is used for the binary stream...

Thanks, yes, using `is` was awful.  I changed that in the replacement PR.

**Comment by KevinRansom** [comment]
Thanks Don,

**Pull request closed without merging**